google-ads-examples/src/main/java/com/google/ads/googleads/examples/misc/NavigateSearchResultPagesCachingTokens.java

```diff
@@ -0,0 +1,247 @@
+   * @param pageTokens the cache of page tokens to use and update.
+   */
+  private static void fetchAndPrintPageResults(
+      GoogleAdsClient googleAdsClient,
+      long customerId,
+      String query,
+      int pageNumber,
+      List<String> pageTokens) {
+    int currentPageNumber;
+    // There is no need to fetch the pages we already know the page tokens for.
+    if (pageTokens.size() >= pageNumber) {
+      System.out.println(
+          "The token of the requested page was cached, we will use it to get the results.");
+      currentPageNumber = pageNumber;
+    } else {
+      System.out.printf(
+          "The token of the requested page was never cached, we will use the closest page we know"
+              + " the token for (page %d) and sequentially get pages from there.%n",
+          pageTokens.size());
+      currentPageNumber = pageTokens.size();
+    }
+
+    // Fetches next pages in sequence and caches their tokens until the requested page results
+    // are returned.
+    while (currentPageNumber <= pageNumber) {
+      // Fetches the next page.
+      System.out.printf("Fetching page %d...%n", currentPageNumber);
+      SearchGoogleAdsRequest request =
+          SearchGoogleAdsRequest.newBuilder()
+              .setCustomerId(Long.toString(customerId))
+              .setPageSize(PAGE_SIZE)
+              .setQuery(query)
+              .setReturnTotalResultsCount(true)
+              // Uses the page token cached for the current page number.
+              .setPageToken(pageTokens.get(currentPageNumber - 1))
+              .build();
+      try (GoogleAdsServiceClient googleAdsServiceClient =
+          googleAdsClient.getLatestVersion().createGoogleAdsServiceClient()) {
+        SearchPagedResponse response = googleAdsServiceClient.search(request);
+        cacheNextPageToken(pageTokens, response.getPage(), currentPageNumber);
+
+        // Prints only the results for the requested page.
+        if (currentPageNumber == pageNumber) {
+          // Prints the results of the requested page.
+          System.out.printf("Printing results found for page %d:%n", pageNumber);
+          for (GoogleAdsRow googleAdsRow : response.getPage().getResponse().getResultsList()) {
+            System.out.printf(
+                "- Campaign with ID %d and name '%s'.%n",
+                googleAdsRow.getCampaign().getId(), googleAdsRow.getCampaign().getName());
+          }
+        }
+
+        currentPageNumber++;
+      }
+    }
+  }
+  // [END navigate_search_result_pages_caching_tokens]
+
+  /**
+   * Updates the cache of page tokens based on a page that was retrieved.
+   *
+   * @param pageTokens the cache of page tokens to update.
+   * @param page the page that was retrieved.
+   * @param pageNumber the number of the page that was retrieved.
+   */
+  private static void cacheNextPageToken(List<String> pageTokens, SearchPage page, int pageNumber) {
```

Comment:
Perhaps it's just me, but it feels a bit strange to pass the `pageNumber` of `page` here instead of the page number corresponding to the token that will be cached.

Reply:
Agreed, and I think this makes more sense in the context of a map.